src/Features/Core/Portable/ExternalAccess/UnitTesting/API/UnitTestingStackTraceService.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.StackTraceExplorer;
+using System.Collections.Immutable;
+using System;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.API
+{
+    internal sealed class UnitTestingStackTraceService
+    {
+        private readonly IStackTraceExplorerService _stackTraceService;
+
+        public UnitTestingStackTraceService(HostWorkspaceServices services)
+            => _stackTraceService = services.GetRequiredService<IStackTraceExplorerService>();
+
+        public async Task<Frame?> TryParseAsync(string line, CancellationToken cancellationToken)
+        {
+            var result = await StackTraceAnalyzer.AnalyzeAsync(line, cancellationToken).ConfigureAwait(false);
+
+            if (result.ParsedFrames.Length == 0)
+            {
+                return null;
+            }
+
+            if (result.ParsedFrames.Length > 1)
+            {
+                throw new InvalidOperationException();
+            }
+
+            return new Frame(result.ParsedFrames.Single(), _stackTraceService);
+        }
+
+        public readonly struct Frame
+        {
+            private readonly ParsedFrame _parsedFrame;
+            private readonly IStackTraceExplorerService _stackTraceService;
+
+            public Frame(ParsedFrame parsedFrame, IStackTraceExplorerService service)
```

Comment:
Should the constructor be internal instead?

Reply:
hmm... probably :)

Reply:
Ah I was thinking that the constructor needn't be exposed to UT. But I guess that is not possible since this is not a nested class. So its ok to ignore this (and the below) comment.

Reply:
Actually wait :) this is a nested struct - so it would be nice to make this private since nothing else needs to construct it :) @ryzngard

Reply:
They can't be private because then the API couldn't construct them. The only other option is to make an interface and have the implementing type private. Not sure it's worth the trouble